apps/prairielearn/src/lib/db-types.ts

```diff
@@ -235,6 +235,29 @@ export const QuestionSchema = z.object({
 });
 export type Question = z.infer<typeof QuestionSchema>;
 
+export const WorkspaceSchema = z.object({
+  created_at: DateFromISOString,
+  heartbeat_at: DateFromISOString.nullable(),
+  hostname: z.string().nullable(),
+  id: IdSchema,
+  launch_port: IdSchema.nullable(), // TODO: this should be a number
```

Comment:
This is a pending discussion. Should the type be changed? Same for version.

Reply:
I can't see any reason `launch_port` should be a `BIGINT` given that port numbers can't exceed 65535 anyways; I think `version` could quite safely be an integer as well. In both cases, I'd be reasonably comfortable having Zod treat them as a plain number instead of a stringified bigint.

Changing the type would require a 4-phase migration (add new column and start writing to it; run a batched migration to rewrite all existing rows; stop reading from the old column; drop the old column). Maybe there's value in that, but it seems like a low priority and low benefit change.

Reply:
My general philosophy here is that the TypeScript type should be a faithful mapping of the postgres type. In this case `workspaces.launch_port` is a `bigint` so I think an IdSchema is a reasonable choice.

However, TCP ports actually max out at 65353, right? So the postgres type should be an `int` and the TypeScript type a Number. I'd be on board with changing these, but I don't feel strongly about it -- I don't think this will actually improve anything very much on in our code?

The same applies to `version`, although in this case I think that there is a better argument for changing it because in the code we actually want to treat this as a number.

Reply:
So should I use `z.coerce.number()`?

Reply:
I'd be on board with `z.coerce.number()` so long as we also do the migrations to change the type to `integer` in postgres once all the code is converted. @nwalters512 I believe we can avoid the 4-phase migration if we:
1. First convert all code to use `z.coerce.number()`
2. Run a migration to change the column type to `integer`
Am I thinking about this correctly?

Reply:
The migration to change the column type requires a table rewrite, which I'd guess would be pretty slow. I think @nwalters512 mentioned the multi-stage migration so that this rewrite can be done in the background.

Reply:
Yep, that's exactly it - we want to avoid locking table rewrites whenever possible, especially for a table as big as this one.

Reply:
Should the first stage of this migration be in this PR or in a new one?

Reply:
Created an issue (#9097) for this. This PR can focus on modernizing the code. This is now ready for review.